skfda/preprocessing/dim_reduction/_pace.py

```diff
@@ -0,0 +1,1441 @@
+        >>> from skfda.representation import FDataIrregular
+        >>> from skfda.preprocessing.dim_reduction import PACE
+
+        >>> points = np.array([0.0, 1.0, 0.0, 1.0])
+        >>> values = np.array([1.0, 0.0, 0.0, 2.0])
+        >>> start_indices = np.array([0, 2])
+
+        >>> fd = FDataIrregular(
+        ...     points=points,
+        ...     values=values,
+        ...     start_indices=start_indices,
+        ... )
+
+        >>> pace = PACE(
+        ...     n_components=2,
+        ...     bandwidth_mean=np.array([0.1, 10]),
+        ...     bandwidth_cov=np.array([0.1, 10]),
+        ... )
+        >>> scores = pace.fit_transform(fd)
+        >>> expected = np.array([
+        ...     [-0.04886943, -0.00023432],
+        ...     [ 0.04886943,  0.00023432],
+        ... ])
+        >>> np.allclose(scores, expected)
+        True
+        >>> round(float(pace.sigma2_), 3)
+        0.973
+
+    References:
+        .. footbibliography::
+    """
+
+    def _check_bandwidth(
+        self,
+        bandwidth: float | NDArrayFloat,
+    ) -> tuple[float | None, tuple[float, float] | None]:
+        """
+        Check if the bandwidth has the correct form.
+
+        Args:
+            bandwidth: Bandwidth to check.
+
+        Returns:
+            A 2-element tuple with the value if the bandwidth is a float, or
+            None otherwise, and the search range if the bandwidth is a tuple,
+            or None otherwise. In the case that the bandwidth is None, the
+            function returns None, (0.1, 10.0) as the default search range.
+        """
+        if isinstance(bandwidth, float) and bandwidth <= 0:
+            error_msg = "Given bandwidth values must be positive."
+            raise ValueError(error_msg)
+
+        tuple_length = 2
+
+        if isinstance(bandwidth, Sequence) and (
+            len(bandwidth) != tuple_length or (
+                not all(isinstance(b, (float, int)) for b in bandwidth)
+            ) or bandwidth[0] <= 0 or bandwidth[1] <= bandwidth[0]
+        ):
+            error_msg = (
+                "Bandwidth search ranges must be a non-decreasing 2-sequence "
+                "of floats."
+            )
+            raise ValueError(error_msg)
+
+        if isinstance(bandwidth, float):
+            return bandwidth, None
+
+        return None, (bandwidth[0], bandwidth[1])
+
+    def __init__(  # noqa: PLR0913
+        self,
+        *,
+        n_components: float | None = None,
+        assume_noisy: bool = True,
+        kernel_mean: KernelFunction = gaussian_kernel,
+        bandwidth_mean: float | NDArrayFloat,
+        kernel_cov: KernelFunction = gaussian_kernel,
+        bandwidth_cov: float | NDArrayFloat,
+        bw_cov_n_grid_points: int = 30,
+        n_grid_points: int = 51,
+        boundary_effect_interval: Sequence[float] = (0.0, 1.0),
+        variance_error_interval: Sequence[float] = (0.25, 0.75),
+    ) -> None:
+        if n_components is None:
```

Comment:
In scikit-learn estimators, the `__init__` should only store the passed parameters as attributes. Validation is delayed until `fit`. This is because `__init__` is not the only way to set parameters: estimators are usually cloned and their parameters are changed during hyperparameter selection.